runtime/gc_include/ObjectAllocationAPI.hpp

```diff
@@ -243,26 +253,32 @@ class MM_ObjectAllocationAPI
 				}
 
 				/* Initialize the object */
-				objectHeader->clazz = (j9objectclass_t)(UDATA)arrayClass;
-				objectHeader->size = size;
-
-				if (initializeSlots) {
-					memset(objectHeader + 1, 0, dataSize);
+				if (J9VMTHREAD_COMPRESS_OBJECT_REFERENCES(currentThread)) {
+					J9IndexableObjectContiguousCompressed *header = (J9IndexableObjectContiguousCompressed*)objectHeader;
+					header->clazz = (U_32)(UDATA)arrayClass;
+					header->size = size;
+					if (initializeSlots) {
+						memset(header + 1, 0, dataSize);
+					}
+				} else {
+					J9IndexableObjectContiguousFull *header = (J9IndexableObjectContiguousFull*)objectHeader;
+					header->clazz = (UDATA)arrayClass;
+					header->size = size;
+					if (initializeSlots) {
+						memset(header + 1, 0, dataSize);
+					}
 				}
 				if (memoryBarrier) {
 					VM_AtomicSupport::writeBarrier();
 				}
-				instance = (j9object_t) objectHeader;
+				instance = objectHeader;
 			}
 		} else {
-			/* Discontiguous Array */
-#if J9SIZEOF_J9IndexableObjectDiscontiguous <= J9_GC_MINIMUM_OBJECT_SIZE
-
-			/* Calculate the object size */
+			/* Zero-length array is discontiguous - assume minimum object size */
 			UDATA allocateSize = J9_GC_MINIMUM_OBJECT_SIZE;
 
 			/* Allocate the memory */
-			J9IndexableObjectDiscontiguous *objectHeader = NULL;
+			void *objectHeader = NULL;
```

Comment:
I would cast here to `j9object_t `

Reply:
I dont understand the comment .

Reply:
`j9object_t  objectHeader = NULL;` as in `J9IndexableObjectContiguous` initialization

Reply:
I could do that, but I'm using it as a pointer type (of two different types) shortly after, so casting to the final type at the end seemed more sensible.

Reply:
the final cast is `instance = (j9object_t)objectHeader;`

Reply:
Yes, but if I change it, it's two casts from the random type to pointers, instead of one cast to the random type at the end.

Reply:
I'll change it if you like (on the assumption that j9object_t will always be castable to a pointer type).

Reply:
I have made the change (I realized it was inconsistent with the mixed allocation path).